crucible-llvm/src/Lang/Crucible/LLVM/MemModel/Generic.hs

```diff
@@ -1135,29 +1135,43 @@ isAllocatedMut mutOk sym w minAlign (llvmPointerView -> (blk, off)) sz m =
     -- @inThisAllocation a allocatedSz@ produces the predicate that
     -- records whether the pointer @ptr@ of size @sz@ falls within the
     -- allocation of block @a@ of size @allocatedSz@.
-    inThisAllocation :: forall w'. Natural -> Maybe (SymBV sym w') -> IO (Pred sym)
-    inThisAllocation a Nothing = isSameBlock sym blk a
-      -- If the allocation is unbounded, we just have to check we are in the right block
-    inThisAllocation a (Just allocSize)
+    inThisAllocation :: forall w'. Maybe (SymBV sym w') -> IO (Pred sym)
+    inThisAllocation Nothing =
+      case sz of
+        Nothing ->
+          -- Unbounded access of an unbounded allocation must start at offset 0.
+          bvEq sym off =<< bvLit sym w 0
+        Just currSize ->
+          -- Bounded access of an unbounded allocation requires
+          -- that @offset + size <= 2^w@.
+          do zeroSize <- bvEq sym currSize =<< bvLit sym w 0
+             noWrap <- bvUle sym off =<< bvNeg sym currSize
+             orPred sym zeroSize noWrap
```

Comment:
Can you explain why this is the right check? It isn't obvious that this implements the inequality from the comment..

Reply:
This is checking `offset <= bvNeg size`, which means `offset <= 2^w - size` for nonzero `size`. I arranged it this way so that it would simplify well in the common case where `size` is concrete. Should I add all that to the comment?

Reply:
Oh, I see.  Yeah, just adding that little bit of algebra to the comment would help, I think.